README.md
# tm-load-test

`tm-load-test` is a distributed load testing tool (and framework) for load
testing [Tendermint](https://tendermint.com/) networks and aims to effectively
be the successor to [`tm-bench`](https://github.com/tendermint/tendermint/tree/v0.32.x/tools/tm-bench).

Naturally, any  transactions sent to a Tendermint network are specific to the
ABCI application running on that network. As such, the `tm-load-test` tool comes
with built-in support for the `kvstore` ABCI application, but you can [build
your own clients](./pkg/loadtest/README.md) for your own apps.

## Requirements

`tm-load-test` is currently tested using Go v1.18.

## Building

To build the `tm-load-test` binary in the `build` directory:

```bash
make
```

## Usage

`tm-load-test` can be executed in one of two modes: **standalone**, or
**coordinator/worker**.

### Standalone Mode

In standalone mode, `tm-load-test` operates in a similar way to `tm-bench`:

```bash
tm-load-test -c 1 -T 10 -r 1000 -s 250 \
    --broadcast-tx-method async \
    --endpoints ws://tm-endpoint1.somewhere.com:26657/websocket,ws://tm-endpoint2.somewhere.com:26657/websocket
```

To see a description of what all of the parameters mean, simply run:

```bash
tm-load-test --help
```

### Coordinator/Worker Mode

In coordinator/worker mode, which is best used for large-scale, distributed load
testing, `tm-load-test` allows you to have multiple worker machines connect to a
single coordinator to obtain their configuration and coordinate their operation.

The coordinator acts as a simple WebSockets host, and the workers are WebSockets
clients.

On the coordinator machine:

```bash
# Run tm-load-test with similar parameters to the standalone mode, but now
# specifying the number of workers to expect (--expect-workers) and the host:port
# to which to bind (--bind) and listen for incoming worker requests.
tm-load-test \
    coordinator \
    --expect-workers 2 \
    --bind localhost:26670 \
    -c 1 -T 10 -r 1000 -s 250 \
    --broadcast-tx-method async \
    --endpoints ws://tm-endpoint1.somewhere.com:26657/websocket,ws://tm-endpoint2.somewhere.com:26657/websocket
```

On each worker machine:

```bash
# Just tell the worker where to find the coordinator - it will figure out the rest.
tm-load-test worker --coordinator localhost:26680
```

For more help, see the command line parameters' descriptions:

```bash
tm-load-test coordinator --help
tm-load-test worker --help
```

### Endpoint Selection Strategies

As of v0.5.1, an endpoint selection strategy can now be given to `tm-load-test`
as a parameter (`--endpoint-select-method`) to control the way in which
endpoints are selected for load testing. There are several options:

1. `supplied` (the default) - only use the supplied endpoints (via the
   `--endpoints` parameter) to submit transactions.
2. `discovered` - only use endpoints discovered through the supplied endpoints
   (by way of crawling the Tendermint peers' network info), but do not use any
   of the supplied endpoints.
3. `any` - use both the supplied and discovered endpoints to perform load
   testing.

**NOTE**: These selection strategies only apply if, and only if, the
`--expect-peers` parameter is supplied and is non-zero. The default behaviour if
`--expect-peers` is not supplied is effectively the `supplied` endpoint
selection strategy.

### Minimum Peer Connectivity

As of v0.6.0, `tm-load-test` can now wait for a minimum level of P2P
connectivity before starting the load testing. By using the
`--min-peer-connectivity` command line switch, along with `--expect-peers`, one
can restrict this.

What this does under the hood is that it checks how many peers are in each
queried peer's address book, and for all reachable peers it checks what the
minimum address book size is. Once the minimum address book size reaches the
configured value, the load testing can begin.

### Customizing

To implement your own client type to load test your own Tendermint ABCI
application, see the [`loadtest` package docs here](./pkg/loadtest/README.md).

## Monitoring

As of v0.4.1, `tm-load-test` exposes a number of metrics when in coordinator/worker
mode, but only from the coordinator's web server at the `/metrics` endpoint. So if
you bind your coordinator node to `localhost:26670`, you should be able to get these
metrics from:

```bash
curl http://localhost:26670/metrics
```

The following kinds of metrics are made available here:

* Total number of transactions recorded from the coordinator's perspective
  (across all workers)
* Total number of transactions sent by each worker
* The status of the coordinator node, which is a gauge that indicates one of the
  following codes:
  * 0 = Coordinator starting
  * 1 = Coordinator waiting for all peers to connect
  * 2 = Coordinator waiting for all workers to connect
  * 3 = Load test underway
  * 4 = Coordinator and/or one or more worker(s) failed
  * 5 = All workers completed load testing successfully
* The status of each worker node, which is also a gauge that indicates one of
  the following codes:
  * 0 = Worker connected
  * 1 = Worker accepted
  * 2 = Worker rejected
  * 3 = Load testing underway
  * 4 = Worker failed
  * 5 = Worker completed load testing successfully
* Standard Prometheus-provided metrics about the garbage collector in
  `tm-load-test`
* The ID of the load test currently underway (defaults to 0), set by way of the
  `--load-test-id` flag on the coordinator

## Aggregate Statistics

As of `tm-load-test` v0.7.0, one can now write simple aggregate statistics to a
CSV file once testing completes by specifying the `--stats-output` flag:

```bash
# In standalone mode
tm-load-test -c 1 -T 10 -r 1000 -s 250 \
    --broadcast-tx-method async \
    --endpoints ws://tm-endpoint1.somewhere.com:26657/websocket,ws://tm-endpoint2.somewhere.com:26657/websocket \
    --stats-output /path/to/save/stats.csv

# From the coordinator in coordinator/worker mode
tm-load-test \
    coordinator \
    --expect-workers 2 \
    --bind localhost:26670 \
    -c 1 -T 10 -r 1000 -s 250 \
    --broadcast-tx-method async \
    --endpoints ws://tm-endpoint1.somewhere.com:26657/websocket,ws://tm-endpoint2.somewhere.com:26657/websocket \
    --stats-output /path/to/save/stats.csv
```

The output CSV file has the following format at present:

```csv
Parameter,Value,Units
total_time,10.002,seconds
total_txs,9000,count
avg_tx_rate,899.818398,transactions per second
```

## Development

To run the linter and the tests:

```bash
make lint
make test
```

### Integration Testing

Integration testing requires Docker to be installed locally.

```bash
make integration-test
```

This integration test:

1. Sets up a 4-validator, fully connected Tendermint Core-based network on a
   192.168.0.0/16 subnet (the same kind of testnet as the Tendermint Core
   localnet).
2. Executes integration tests against the network in series (it's important that
   integration tests be executed in series so as to not overlap with one
   another).
3. Tears down the 4-validator network, reporting code coverage.



CHANGELOG.md
# Changelog

## v1.3.0

*Jan 19th, 2023*

* [\#168](https://github.com/informalsystems/tm-load-test/pull/168) - Remove
  dependency on Tendermint Core, following
  [tendermint/tendermint\#9972](https://github.com/tendermint/tendermint/issues/9972).
  Also improves integration tests.

## v1.2.0

* [\#167](https://github.com/informalsystems/tm-load-test/pull/167) -
  Temporarily replace the Tendermint Core dependency with Informal Systems' fork

## v1.1.0

This minor release just bumps some dependencies.

* [\#163](https://github.com/informalsystems/tm-load-test/pull/163) - Bump
  supported version of Tendermint Core to v0.34.24.

## v1.0.0
* [\#113](https://github.com/informalsystems/tm-load-test/pull/113) - Dropped
  the master/slave terminology. "Master" nodes have now been renamed to
  "Coordinator", and "Slave" nodes have been renamed to "Worker".

## v0.9.0
* [\#47](https://github.com/informalsystems/tm-load-test/pull/47) - Makes sure
  that the KVStore client's `GenerateTx` method honours the preconfigured
  transaction size. **NB: This involves a breaking API change!** Please see
  [the client API documentation](./pkg/loadtest/README.md) for more details.

## v0.8.0
* [\#42](https://github.com/informalsystems/tm-load-test/pull/42) - Add Prometheus
  gauge for when load test is underway. This indicator exposes a customizable
  load test ID.

## v0.7.1
* Re-released due to v0.7.0 being incorrectly tagged

## v0.7.0
* [\#39](https://github.com/informalsystems/tm-load-test/pull/40) - Add basic
  aggregate statistics output to CSV file.
* Added integration test for standalone execution happy path.

## v0.6.2
* [\#37](https://github.com/informalsystems/tm-load-test/pull/37) - Fix average
  transaction throughput rate in reporting/metrics.

## v0.6.1
* [\#35](https://github.com/informalsystems/tm-load-test/pull/35) - Minor fix for
  broken shutdown-wait functionality.

## v0.6.0
* [\#33](https://github.com/informalsystems/tm-load-test/pull/33) - Add ability
  to wait for a minimum level of connectivity between peers before starting the
  load testing (through a `--min-peer-connectivity` command line switch).

## v0.5.1
* [\#31](https://github.com/informalsystems/tm-load-test/pull/31) - Expand on
  endpoint selection strategy to now allow for 3 different strategies:
  `supplied`, `discovered` and `any`. Allows for specifying of a seed node
  endpoint that one doesn't want to use during the actual load testing.

## v0.5.0
* [\#23](https://github.com/informalsystems/tm-load-test/pull/23) - Add
  feature to wait for Tendermint network stabilization before starting load
  testing (in standalone and master/slave modes).
* [\#28](https://github.com/informalsystems/tm-load-test/pull/28) - Expose the
  tx throughput rates via Prometheus.

## v0.4.2
* Adds version sub-command to the CLI

## v0.4.1
* [\#21](https://github.com/informalsystems/tm-load-test/pull/21) - Add support
  for exposing Prometheus-compatible metrics from the MASTER web server via
  the `/metrics` endpoint. This now provides simple high-level information
  about the overall load test, like number of transactions sent (overall, and
  per-slave), and the state of the master and each attached slave.

## v0.4.0
* [\#20](https://github.com/informalsystems/tm-load-test/pull/20) - Significant
  refactor of the underlying codebase to radically simplify the code and make
  its usage easier and more closely aligned with `tm-bench`.

## v0.3.0
* [\#15](https://github.com/informalsystems/tm-load-test/pull/14) - Add example
  configuration for outage simulation while load testing.
* [\#14](https://github.com/informalsystems/tm-load-test/pull/14) - Add support for
  the outage simulator from the master node during load testing.
* [\#13](https://github.com/informalsystems/tm-load-test/pull/13) - Add basic HTTP
  authentication to `tm-outage-sim-server` utility.
* [\#12](https://github.com/informalsystems/tm-load-test/pull/12) - Add standalone
  mode to be able to run load testing tool locally in a simple way.
* [\#11](https://github.com/informalsystems/tm-load-test/pull/11) - Allow for basic
  HTTP authentication between the master and slave nodes for additional
  security.
* [\#10](https://github.com/informalsystems/tm-load-test/pull/10) - Allow clients
  to continuously send transactions until the maximum time limit is reached
  without having a limit imposed on the number of transactions.

## v0.2.0
* First alpha release.
* Refactored comms mechanism between master and slaves to use HTTP for more
  robust communication and to simplify.
* Added the WebSockets-based client for interacting with Tendermint nodes over
  WebSockets (`kvstore-websockets`).

## v0.1.0
* Initial release to configure release management.



pkg/loadtest/cli.go
package loadtest

import (
	"fmt"
	"os"
	"os/signal"
	"syscall"

	"github.com/informalsystems/tm-load-test/internal/logging"
	"github.com/sirupsen/logrus"
	"github.com/spf13/cobra"
)

// CLIVersion must be manually updated as new versions are released.
const CLIVersion = "v1.3.0"

// cliVersionCommitID must be set through linker settings. See
// https://stackoverflow.com/a/11355611/1156132 for details.
var cliVersionCommitID string

// CLIConfig allows developers to customize their own load testing tool.
type CLIConfig struct {
	AppName              string
	AppShortDesc         string
	AppLongDesc          string
	DefaultClientFactory string
}

var flagVerbose bool

func buildCLI(cli *CLIConfig, logger logging.Logger) *cobra.Command {
	cobra.OnInitialize(func() { initLogLevel(logger) })
	var cfg Config
	rootCmd := &cobra.Command{
		Use:   cli.AppName,
		Short: cli.AppShortDesc,
		Long:  cli.AppLongDesc,
		Run: func(cmd *cobra.Command, args []string) {
			logger.Debug(fmt.Sprintf("Configuration: %s", cfg.ToJSON()))
			if err := cfg.Validate(); err != nil {
				logger.Error(err.Error())
				os.Exit(1)
			}

			if err := ExecuteStandalone(cfg); err != nil {
				os.Exit(1)
			}
		},
	}
	rootCmd.PersistentFlags().StringVar(&cfg.ClientFactory, "client-factory", cli.DefaultClientFactory, "The identifier of the client factory to use for generating load testing transactions")
	rootCmd.PersistentFlags().IntVarP(&cfg.Connections, "connections", "c", 1, "The number of connections to open to each endpoint simultaneously")
	rootCmd.PersistentFlags().IntVarP(&cfg.Time, "time", "T", 60, "The duration (in seconds) for which to handle the load test")
	rootCmd.PersistentFlags().IntVarP(&cfg.SendPeriod, "send-period", "p", 1, "The period (in seconds) at which to send batches of transactions")
	rootCmd.PersistentFlags().IntVarP(&cfg.Rate, "rate", "r", 1000, "The number of transactions to generate each second on each connection, to each endpoint")
	rootCmd.PersistentFlags().IntVarP(&cfg.Size, "size", "s", 250, "The size of each transaction, in bytes - must be greater than 40")
	rootCmd.PersistentFlags().IntVarP(&cfg.Count, "count", "N", -1, "The maximum number of transactions to send - set to -1 to turn off this limit")
	rootCmd.PersistentFlags().StringVar(&cfg.BroadcastTxMethod, "broadcast-tx-method", "async", "The broadcast_tx method to use when submitting transactions - can be async, sync or commit")
	rootCmd.PersistentFlags().StringSliceVar(&cfg.Endpoints, "endpoints", []string{}, "A comma-separated list of URLs indicating Tendermint WebSockets RPC endpoints to which to connect")
	rootCmd.PersistentFlags().StringVar(&cfg.EndpointSelectMethod, "endpoint-select-method", SelectSuppliedEndpoints, "The method by which to select endpoints")
	rootCmd.PersistentFlags().IntVar(&cfg.ExpectPeers, "expect-peers", 0, "The minimum number of peers to expect when crawling the P2P network from the specified endpoint(s) prior to waiting for workers to connect")
	rootCmd.PersistentFlags().IntVar(&cfg.MaxEndpoints, "max-endpoints", 0, "The maximum number of endpoints to use for testing, where 0 means unlimited")
	rootCmd.PersistentFlags().IntVar(&cfg.PeerConnectTimeout, "peer-connect-timeout", 600, "The number of seconds to wait for all required peers to connect if expect-peers > 0")
	rootCmd.PersistentFlags().IntVar(&cfg.MinConnectivity, "min-peer-connectivity", 0, "The minimum number of peers to which each peer must be connected before starting the load test")
	rootCmd.PersistentFlags().StringVar(&cfg.StatsOutputFile, "stats-output", "", "Where to store aggregate statistics (in CSV format) for the load test")
	rootCmd.PersistentFlags().BoolVarP(&flagVerbose, "verbose", "v", false, "Increase output logging verbosity to DEBUG level")

	var coordCfg CoordinatorConfig
	coordCmd := &cobra.Command{
		Use:   "coordinator",
		Short: "Start load test application in COORDINATOR mode",
		Run: func(cmd *cobra.Command, args []string) {
			logger.Debug(fmt.Sprintf("Configuration: %s", cfg.ToJSON()))
			logger.Debug(fmt.Sprintf("Coordinator configuration: %s", coordCfg.ToJSON()))
			if err := cfg.Validate(); err != nil {
				logger.Error(err.Error())
				os.Exit(1)
			}
			if err := coordCfg.Validate(); err != nil {
				logger.Error(err.Error())
				os.Exit(1)
			}
			coord := NewCoordinator(&cfg, &coordCfg)
			if err := coord.Run(); err != nil {
				os.Exit(1)
			}
		},
	}
	coordCmd.PersistentFlags().StringVar(&coordCfg.BindAddr, "bind", "localhost:26670", "A host:port combination to which to bind the coordinator on which to listen for worker connections")
	coordCmd.PersistentFlags().IntVar(&coordCfg.ExpectWorkers, "expect-workers", 2, "The number of workers to expect to connect to the coordinator before starting load testing")
	coordCmd.PersistentFlags().IntVar(&coordCfg.WorkerConnectTimeout, "connect-timeout", 180, "The maximum number of seconds to wait for all workers to connect")
	coordCmd.PersistentFlags().IntVar(&coordCfg.ShutdownWait, "shutdown-wait", 0, "The number of seconds to wait after testing completes prior to shutting down the web server")
	coordCmd.PersistentFlags().IntVar(&coordCfg.LoadTestID, "load-test-id", 0, "The ID of the load test currently underway")

	var workerCfg WorkerConfig
	workerCmd := &cobra.Command{
		Use:   "worker",
		Short: "Start load test application in WORKER mode",
		Run: func(cmd *cobra.Command, args []string) {
			logger.Debug(fmt.Sprintf("Worker configuration: %s", workerCfg.ToJSON()))
			if err := workerCfg.Validate(); err != nil {
				logger.Error(err.Error())
				os.Exit(1)
			}
			worker, err := NewWorker(&workerCfg)
			if err != nil {
				logger.Error("Failed to create new worker", "err", err)
				os.Exit(1)
			}
			if err := worker.Run(); err != nil {
				os.Exit(1)
			}
		},
	}
	workerCmd.PersistentFlags().StringVar(&workerCfg.ID, "id", "", "An optional unique ID for this worker. Will show up in metrics and logs. If not specified, a UUID will be generated.")
	workerCmd.PersistentFlags().StringVar(&workerCfg.CoordAddr, "coordinator", "ws://localhost:26670", "The WebSockets URL on which to find the coordinator node")
	workerCmd.PersistentFlags().IntVar(&workerCfg.CoordConnectTimeout, "connect-timeout", 180, "The maximum number of seconds to keep trying to connect to the coordinator")

	versionCmd := &cobra.Command{
		Use:   "version",
		Short: "Display the version of tm-load-test and exit",
		Run: func(cmd *cobra.Command, args []string) {
			version := CLIVersion
			if len(cliVersionCommitID) > 0 {
				version = fmt.Sprintf("%s-%s", version, cliVersionCommitID)
			}
			fmt.Println("tm-load-test", version)
		},
	}

	rootCmd.AddCommand(coordCmd)
	rootCmd.AddCommand(workerCmd)
	rootCmd.AddCommand(versionCmd)
	return rootCmd
}

func initLogLevel(logger logging.Logger) {
	if flagVerbose {
		logrus.SetLevel(logrus.DebugLevel)
		logger.Debug("Set logging level to DEBUG")
	}
}

// Run must be executed from your `main` function in your Go code. This can be
// used to fast-track the construction of your own load testing tool for your
// Tendermint ABCI application.
func Run(cli *CLIConfig) {
	logger := logging.NewLogrusLogger("main")
	if err := buildCLI(cli, logger).Execute(); err != nil {
		logger.Error("Error", "err", err)
	}
}

func trapInterrupts(onKill func(), logger logging.Logger) chan struct{} {
	sigc := make(chan os.Signal, 1)
	cancelTrap := make(chan struct{})
	signal.Notify(sigc, os.Interrupt, syscall.SIGTERM)
	go func() {
		select {
		case <-sigc:
			logger.Info("Caught kill signal")
			onKill()
		case <-cancelTrap:
			logger.Debug("Interrupt trap cancelled")
		}
	}()
	return cancelTrap
}
